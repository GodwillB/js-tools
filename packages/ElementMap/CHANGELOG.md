# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.0.5](https://github.com/onaio/js-tools/compare/@onaio/element-map@0.0.2...@onaio/element-map@0.0.5) (2019-05-01)

**Note:** Version bump only for package @onaio/element-map

## [0.0.3](https://github.com/onaio/js-tools/compare/@onaio/element-map@0.0.2...@onaio/element-map@0.0.3) (2019-05-01)

**Note:** Version bump only for package @onaio/element-map

## [0.0.2](https://github.com/onaio/js-tools/compare/@onaio/element-map@0.0.1...@onaio/element-map@0.0.2) (2019-03-21)

**Note:** Version bump only for package @onaio/element-map

## 0.0.1 (2019-03-13)

**Note:** Version bump only for package @onaio/element-map
