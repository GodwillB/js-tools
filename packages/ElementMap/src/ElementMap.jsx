// This component takes a list of elements and returns them wrapped
// in the provided HTML tag
import React, { Component } from 'react';
import PropTypes from 'prop-types';

class ElementMap extends Component {
  render() {
    const { items } = this.props;
    const { className } = this.props;
    const { HTMLTag } = this.props;
    // use React.createElement to create the html tag
    // in a dynamic way
    const listItems = items.map((item, key) =>
      React.createElement(
        HTMLTag,
        /* eslint-disable react/no-array-index-key */
        { key, className },
        /* eslint-enable react/no-array-index-key */
        item
      )
    );

    return listItems;
  }
}

ElementMap.propTypes = {
  items: PropTypes.arrayOf(PropTypes.node).isRequired,
  HTMLTag: PropTypes.string.isRequired,
  className: PropTypes.string
};

ElementMap.defaultProps = {
  className: ''
};

export default ElementMap;
