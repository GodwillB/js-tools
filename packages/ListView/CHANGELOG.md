# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.0.3](https://github.com/onaio/js-tools/compare/@onaio/list-view@0.0.2...@onaio/list-view@0.0.3) (2020-03-10)

**Note:** Version bump only for package @onaio/list-view

## [0.0.2](https://github.com/onaio/js-tools/compare/@onaio/list-view@0.0.1...@onaio/list-view@0.0.2) (2019-08-20)

**Note:** Version bump only for package @onaio/list-view

## 0.0.1 (2019-05-31)

**Note:** Version bump only for package @onaio/list-view
