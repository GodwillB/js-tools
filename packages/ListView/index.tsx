import ElementMap from '@onaio/element-map';
import React from 'react';

/** Type definition for renderHeaderFunc */
export type renderHeadersFuncType = (
  items?: React.ReactNode[],
  theadClass?: string,
  thClass?: string,
  trClass?: string
) => Element | null;

/** Type definition for renderRowsFunc */

export type renderRowsFuncType = (
  rowData: React.ReactNode[][],
  tbClass?: string,
  tdClass?: string,
  trClass?: string
) => Element;

/** Renders header section of a table */
export function renderHeadersFunc(
  items?: React.ReactNode[],
  theadClass?: string,
  thClass?: string,
  trClass?: string
) {
  if (items) {
    return (
      <thead className={theadClass}>
        <tr className={trClass}>
          <ElementMap items={items} HTMLTag="th" className={thClass} />
        </tr>
      </thead>
    );
  } else {
    return null;
  }
}

/** Renders body section of the table */
export function renderRowsFunc(
  rowData: React.ReactNode[][],
  tbClass?: string,
  trClass?: string,
  tdClass?: string
) {
  const rows = rowData.map((items, itemKey) => (
    <tr key={itemKey} className={trClass}>
      <ElementMap items={items} HTMLTag="td" className={tdClass} />
    </tr>
  ));
  return <tbody className={tbClass}>{rows}</tbody>;
}

/** Interface to define props for ListView */
export interface ListViewProps {
  data: React.ReactNode[][];
  headerItems?: React.ReactNode[];
  tableClass?: string;
  tbodyClass?: string;
  tdClass?: string;
  theadClass?: string;
  thClass?: string;
  trClass?: string;
  renderHeaders?: renderHeadersFuncType;
  renderRows?: renderRowsFuncType;
}

/** A simple component that takes in data and renders it in a tabular view  */
const ListView: React.ElementType = (props: ListViewProps) => {
  const {
    data,
    headerItems,
    renderHeaders,
    renderRows,
    tableClass,
    tdClass,
    tbodyClass,
    theadClass,
    thClass,
    trClass
  } = props;

  return (
    <table className={tableClass}>
      {renderHeaders && renderHeaders(headerItems, theadClass, tdClass, trClass)}
      {renderRows && renderRows(data, tbodyClass, thClass, trClass)}
    </table>
  );
};

ListView.defaultProps = {
  renderHeaders: renderHeadersFunc,
  renderRows: renderRowsFunc,
  tableClass: 'listview',
  tbodyClass: 'listview-tbody',
  tdClass: 'listview-td',
  thClass: 'listview-th',
  theadClass: 'listview-thead',
  trClass: 'listview-primary'
};

export default ListView;
