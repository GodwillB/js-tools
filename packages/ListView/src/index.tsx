import ElementMap from '@onaio/element-map';
import React from 'react';

/** Type definition for renderHeaderFunc */
export type renderHeadersFuncType = (
  items?: React.ReactNode[],
  theadClass?: string,
  thClass?: string,
  trClass?: string
) => JSX.Element | null;

/** Type definition for renderRowsFunc */
export type renderRowsFuncType = (
  rowData: React.ReactNode[][],
  tbClass?: string,
  tdClass?: string,
  trClass?: string
) => JSX.Element;

/** Renders header section of a table */
export function renderHeadersFunc(
  items?: React.ReactNode[],
  theadClass?: string,
  thClass?: string,
  trClass?: string
) {
  if (items) {
    return (
      <thead className={theadClass ? theadClass : ''}>
        <tr className={trClass ? trClass : ''}>
          <ElementMap items={items} HTMLTag="th" className={thClass ? thClass : ''} />
        </tr>
      </thead>
    );
  } else {
    return null;
  }
}

/** Renders body section of the table */
export function renderRowsFunc(
  rowData: React.ReactNode[][],
  tbClass?: string,
  tdClass?: string,
  trClass?: string
) {
  const rows = rowData.map((items, itemKey) => (
    <tr key={itemKey} className={trClass ? trClass : ''}>
      <ElementMap items={items} HTMLTag="td" className={tdClass ? tdClass : ''} />
    </tr>
  ));
  return <tbody className={tbClass ? tbClass : ''}>{rows}</tbody>;
}

/** Interface to define props for ListView */
export interface ListViewProps {
  data: React.ReactNode[][];
  headerItems: React.ReactNode[];
  tableClass: string;
  tbodyClass: string;
  tdClass: string;
  theadClass: string;
  thClass: string;
  trClass: string;
  renderHeaders: renderHeadersFuncType;
  renderRows: renderRowsFuncType;
}

/** default props for ListView */
const defaultListViewProps: Partial<ListViewProps> = {
  renderHeaders: renderHeadersFunc,
  renderRows: renderRowsFunc,
  tableClass: 'listview',
  tbodyClass: 'listview-tbody',
  tdClass: 'listview-td',
  thClass: 'listview-th',
  theadClass: 'listview-thead',
  trClass: 'listview-tr'
};

/** A simple component that takes in data and renders it in a tabular view  */
const ListView = (props: ListViewProps) => {
  const {
    data,
    headerItems,
    renderHeaders,
    renderRows,
    tableClass,
    tdClass,
    tbodyClass,
    theadClass,
    thClass,
    trClass
  } = props;

  return (
    <table className={tableClass}>
      {renderHeaders && renderHeaders(headerItems, theadClass, tdClass, trClass)}
      {renderRows && renderRows(data, tbodyClass, thClass, trClass)}
    </table>
  );
};

ListView.defaultProps = defaultListViewProps;

export default ListView;
