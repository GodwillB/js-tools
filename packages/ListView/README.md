# ListView

A simple re-usable `React` component that lists data items in a table.

## How it works

This component works by taking in data items, optional table classes as props and renders a table with the provided classes. The component allows one to overide the methods rendering the table headers and table rows by passing them as props to the component too.

## Installation & Usage

### Examples

```tsx
import ListView from '@onaio/listview';

const props = {
  data: [['Ed', 6, 'Taller'], ['Edd', 12, 'Tallest'], ['Eddie', 17, 'Tall']],
  headerItems: ['Name', 'Age', 'Height'],
  tableClass: 'table-striped',
  tbodyClass: 'table-active',
  tdClass: 'td-primary',
  theadClass: 'thead-dark',
  thClass: 'th-primary',
  trClass: 'tr-warning'
};
<ListView {...props} />;

//output should now look like:

<table class="table-striped">
  <thead class="thead-dark">
    <tr class="tr-warning">
      <th className="th-primary" key="0">
        {' '}
        Name{' '}
      </th>
      <th className="th-primary" key="1">
        {' '}
        Age{' '}
      </th>
      <th className="th-primary" key="2">
        {' '}
        Height{' '}
      </th>
    </tr>
  </thead>
  <tbody className="table-active">
    <tr class="tr-warning" key="0">
      <td className="td-primary" key="0">
        {' '}
        Ed{' '}
      </td>
      <td className="td-primary" key="1">
        {' '}
        6{' '}
      </td>
      <td className="td-primary" key="2">
        {' '}
        Taller{' '}
      </td>
    </tr>
    <tr class="tr-warning" key="1">
      <td className="td-primary" key="0">
        {' '}
        Edd{' '}
      </td>
      <td className="td-primary" key="1">
        {' '}
        12{' '}
      </td>
      <td className="td-primary" key="2">
        {' '}
        Tallest{' '}
      </td>
    </tr>
    <tr class="tr-warning" key="2">
      <td className="td-primary" key="0">
        {' '}
        Eddie{' '}
      </td>
      <td className="td-primary" key="1">
        {' '}
        17{' '}
      </td>
      <td className="td-primary" key="2">
        {' '}
        Tall{' '}
      </td>
    </tr>
  </tbody>
</table>;
```

## Props

The ListView component takes these props:

**data** - prop that holds data items that will be rendered on the table e.g. `[['Ed', 6, 'Taller'], ['Edd', 12, 'Tallest'], ['Eddie', 17, 'Tall']]`

**headerItems** - prop that holds header items of the table an example would be `['Name', 'Age', 'Height']`

**tableClass** - prop that holds table element class e.g `table-striped`

**tbodyClass** - prop that holds table body element class e.g. `table-active`

**tdClass** - prop that holds table data element class e.g. `td-primary`

**theadClass** - prop that holds table head element class e.g. `thead-dark`

**thClass** - prop that holds table header element class e.g. `th-primary`

**trClass** - prop that holds table header element class e.g. `tr-warning`

**renderHeaders** - This prop is a method that takes header data items and classes to render the header section of our table. The method can be passed as a prop into the component to overide the existing method. The current implementation uses `ElementMap` component to loop through the data items adding them inside th elements.
e.g.

```renderHeaders: (items, cssClass) => (
      <thead className={cssClass}>
        <tr>
          <th colSpan={2}>Top Header</th>
        </tr>
        <tr>
          <ElementMap items={items} HTMLTag="th" />
        </tr>
      </thead>
    )
```

**renderRows** - Very simmillar to renderHeaders prop this is a method that renders the table body. The method takes row data items and css classes as parameters, `ElementMap` component is also used in this method to loop through the data items adding them inside td elements.

## props with sample values

```tsx
const props = {
  data: [['Ed', 6, 'Taller'], ['Edd', 12, 'Tallest'], ['Eddie', 17, 'Tall']],
  headerItems: ['Name', 'Age', 'Height'],
  tableClass: 'table-striped',
  tbodyClass: 'table-active',
  tdClass: 'td-primary',
  theadClass: 'thead-dark',
  thClass: 'th-primary',
  trClass: 'tr-warning',

  renderHeaders: (items, theadClass, thClass, trClass) => {
   return (
     <thead className={theadClass}>
      <tr className={trClass}>
        <ElementMap items={items} HTMLTag="th" className={thClass} />
      </tr>
    </thead>);
  }
  renderRows: (rowData, tbodyClass, tdClass, trClass) => {
    const rows = rowData.map((items, itemKey) => (
      <tr key={itemKey} className={trClass}>
        <ElementMap items={items} HTMLTag="td" className={tdClass} />
      </tr>
    ));
    return <tbody className={tbodyClass}>{rows}</tbody>;
  }
};
```
