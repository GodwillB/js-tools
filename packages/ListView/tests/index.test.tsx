import ElementMap from '@onaio/element-map';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import ListView, { ListViewProps } from '../';

describe('ListView', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    const props = {
      data: [['Data 1'], ['Data 2'], ['Data 3']],
      headerItems: ['Columns'],
      tableClass: 'table-striped',
      tbodyClass: 'table-active',
      tdClass: 'table-primary',
      thClass: 'table-info',
      theadClass: 'thead-dark',
      trClass: 'table-success'
    };
    shallow(<ListView {...props} />);
  });

  it('renders correctly', () => {
    const props = {
      data: [['Ed', 6, 'Taller'], ['Edd', 12, 'Tallest'], ['Eddie', 17, 'Tall']],
      headerItems: ['Name', 'Age', 'Height'],
      tableClass: 'table-striped',
      tbodyClass: 'table-active',
      tdClass: 'table-primary',
      thClass: 'table-info',
      theadClass: 'thead-dark',
      trClass: 'table-success'
    };
    const wrapper = mount(<ListView {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly with different types of data', () => {
    const props: ListViewProps = {
      data: [['Data 1'], [999], ['Data 3']],
      headerItems: ['Columns'],
      tableClass: 'table-striped',
      tbodyClass: 'table-active',
      tdClass: 'table-primary',
      thClass: 'table-info',
      theadClass: 'thead-dark',
      trClass: 'table-success'
    };
    const wrapper = mount(<ListView {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly with rows containing components', () => {
    const SomeComponent: React.ElementType = props => {
      return (
        <div>
          <span>{props.value}</span>
        </div>
      );
    };
    const tableProps: ListViewProps = {
      data: [
        ['Data 1'],
        [<div key="123">Data 2</div>],
        [<SomeComponent key="456" value="Data 3" />]
      ],
      headerItems: ['Columns'],
      tableClass: 'table-striped',
      tbodyClass: 'table-active',
      tdClass: 'table-primary',
      thClass: 'table-info',
      theadClass: 'thead-dark',
      trClass: 'table-success'
    };
    const wrapper = mount(<ListView {...tableProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly without thead, tableClass, tbodyClass, tdClass, trClass, thClass', () => {
    const props: ListViewProps = {
      data: [['Data 1'], [999], ['Data 3']]
    };
    const wrapper = mount(<ListView {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly with custom renderHeaders prop', () => {
    const props = {
      data: [['Ed', 6], ['Edd', 12], ['Eddie', 17]],
      headerItems: ['Name', 'Age'],
      renderHeaders: (items, theadClass, thClass, trClass) => (
        <thead className={theadClass}>
          <tr className={trClass}>
            <th colSpan={2}>Top Header</th>
          </tr>
          <tr>
            <ElementMap items={items} HTMLTag="th" className={thClass} />
          </tr>
        </thead>
      )
    };
    const wrapper = mount(<ListView {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly with no renderHeaders prop', () => {
    const props = {
      data: [['Ed'], ['Edd'], ['Eddie']],
      headerItems: ['Name'],
      renderHeaders: undefined
    };
    const wrapper = mount(<ListView {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly with custom renderRows prop', () => {
    const props = {
      data: [['Ed', 6], ['Edd', 12], ['Eddie', 17]],
      headerItems: ['Name', 'Age'],
      renderRows: (rowData, tbodyClass, tdClass, trClass) => {
        const rows = rowData.map((items, itemKey) => (
          <tr key={itemKey} className={trClass}>
            <ElementMap items={items} HTMLTag="td" className={tdClass} />
          </tr>
        ));
        return <tbody className={tbodyClass}>{rows}</tbody>;
      }
    };
    const wrapper = mount(<ListView {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
