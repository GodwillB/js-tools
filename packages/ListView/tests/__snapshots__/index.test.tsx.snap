// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListView renders correctly 1`] = `
<ListView
  data={
    Array [
      Array [
        "Ed",
        6,
        "Taller",
      ],
      Array [
        "Edd",
        12,
        "Tallest",
      ],
      Array [
        "Eddie",
        17,
        "Tall",
      ],
    ]
  }
  headerItems={
    Array [
      "Name",
      "Age",
      "Height",
    ]
  }
  renderHeaders={[Function]}
  renderRows={[Function]}
  tableClass="table-striped"
  tbodyClass="table-active"
  tdClass="table-primary"
  thClass="table-info"
  theadClass="thead-dark"
  trClass="table-success"
>
  <table
    className="table-striped"
  >
    <thead
      className="thead-dark"
    >
      <tr
        className="table-success"
      >
        <ElementMap
          HTMLTag="th"
          className="table-primary"
          items={
            Array [
              "Name",
              "Age",
              "Height",
            ]
          }
        >
          <th
            className="table-primary"
            key="0"
          >
            Name
          </th>
          <th
            className="table-primary"
            key="1"
          >
            Age
          </th>
          <th
            className="table-primary"
            key="2"
          >
            Height
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="table-active"
    >
      <tr
        className="table-success"
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="table-info"
          items={
            Array [
              "Ed",
              6,
              "Taller",
            ]
          }
        >
          <td
            className="table-info"
            key="0"
          >
            Ed
          </td>
          <td
            className="table-info"
            key="1"
          >
            6
          </td>
          <td
            className="table-info"
            key="2"
          >
            Taller
          </td>
        </ElementMap>
      </tr>
      <tr
        className="table-success"
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="table-info"
          items={
            Array [
              "Edd",
              12,
              "Tallest",
            ]
          }
        >
          <td
            className="table-info"
            key="0"
          >
            Edd
          </td>
          <td
            className="table-info"
            key="1"
          >
            12
          </td>
          <td
            className="table-info"
            key="2"
          >
            Tallest
          </td>
        </ElementMap>
      </tr>
      <tr
        className="table-success"
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="table-info"
          items={
            Array [
              "Eddie",
              17,
              "Tall",
            ]
          }
        >
          <td
            className="table-info"
            key="0"
          >
            Eddie
          </td>
          <td
            className="table-info"
            key="1"
          >
            17
          </td>
          <td
            className="table-info"
            key="2"
          >
            Tall
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly with custom renderHeaders prop 1`] = `
<ListView
  data={
    Array [
      Array [
        "Ed",
        6,
      ],
      Array [
        "Edd",
        12,
      ],
      Array [
        "Eddie",
        17,
      ],
    ]
  }
  headerItems={
    Array [
      "Name",
      "Age",
    ]
  }
  renderHeaders={[Function]}
  renderRows={[Function]}
  tableClass="listview"
  tbodyClass="listview-tbody"
  tdClass="listview-td"
  thClass="listview-th"
  theadClass="listview-thead"
  trClass="listview-primary"
>
  <table
    className="listview"
  >
    <thead
      className="listview-thead"
    >
      <tr
        className="listview-primary"
      >
        <th
          colSpan={2}
        >
          Top Header
        </th>
      </tr>
      <tr>
        <ElementMap
          HTMLTag="th"
          className="listview-td"
          items={
            Array [
              "Name",
              "Age",
            ]
          }
        >
          <th
            className="listview-td"
            key="0"
          >
            Name
          </th>
          <th
            className="listview-td"
            key="1"
          >
            Age
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="listview-tbody"
    >
      <tr
        className="listview-primary"
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              "Ed",
              6,
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            Ed
          </td>
          <td
            className="listview-th"
            key="1"
          >
            6
          </td>
        </ElementMap>
      </tr>
      <tr
        className="listview-primary"
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              "Edd",
              12,
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            Edd
          </td>
          <td
            className="listview-th"
            key="1"
          >
            12
          </td>
        </ElementMap>
      </tr>
      <tr
        className="listview-primary"
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              "Eddie",
              17,
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            Eddie
          </td>
          <td
            className="listview-th"
            key="1"
          >
            17
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly with custom renderRows prop 1`] = `
<ListView
  data={
    Array [
      Array [
        "Ed",
        6,
      ],
      Array [
        "Edd",
        12,
      ],
      Array [
        "Eddie",
        17,
      ],
    ]
  }
  headerItems={
    Array [
      "Name",
      "Age",
    ]
  }
  renderHeaders={[Function]}
  renderRows={[Function]}
  tableClass="listview"
  tbodyClass="listview-tbody"
  tdClass="listview-td"
  thClass="listview-th"
  theadClass="listview-thead"
  trClass="listview-primary"
>
  <table
    className="listview"
  >
    <thead
      className="listview-thead"
    >
      <tr
        className="listview-primary"
      >
        <ElementMap
          HTMLTag="th"
          className="listview-td"
          items={
            Array [
              "Name",
              "Age",
            ]
          }
        >
          <th
            className="listview-td"
            key="0"
          >
            Name
          </th>
          <th
            className="listview-td"
            key="1"
          >
            Age
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="listview-tbody"
    >
      <tr
        className="listview-primary"
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              "Ed",
              6,
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            Ed
          </td>
          <td
            className="listview-th"
            key="1"
          >
            6
          </td>
        </ElementMap>
      </tr>
      <tr
        className="listview-primary"
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              "Edd",
              12,
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            Edd
          </td>
          <td
            className="listview-th"
            key="1"
          >
            12
          </td>
        </ElementMap>
      </tr>
      <tr
        className="listview-primary"
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              "Eddie",
              17,
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            Eddie
          </td>
          <td
            className="listview-th"
            key="1"
          >
            17
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly with different types of data 1`] = `
<ListView
  data={
    Array [
      Array [
        "Data 1",
      ],
      Array [
        999,
      ],
      Array [
        "Data 3",
      ],
    ]
  }
  headerItems={
    Array [
      "Columns",
    ]
  }
  renderHeaders={[Function]}
  renderRows={[Function]}
  tableClass="table-striped"
  tbodyClass="table-active"
  tdClass="table-primary"
  thClass="table-info"
  theadClass="thead-dark"
  trClass="table-success"
>
  <table
    className="table-striped"
  >
    <thead
      className="thead-dark"
    >
      <tr
        className="table-success"
      >
        <ElementMap
          HTMLTag="th"
          className="table-primary"
          items={
            Array [
              "Columns",
            ]
          }
        >
          <th
            className="table-primary"
            key="0"
          >
            Columns
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="table-active"
    >
      <tr
        className="table-success"
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="table-info"
          items={
            Array [
              "Data 1",
            ]
          }
        >
          <td
            className="table-info"
            key="0"
          >
            Data 1
          </td>
        </ElementMap>
      </tr>
      <tr
        className="table-success"
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="table-info"
          items={
            Array [
              999,
            ]
          }
        >
          <td
            className="table-info"
            key="0"
          >
            999
          </td>
        </ElementMap>
      </tr>
      <tr
        className="table-success"
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="table-info"
          items={
            Array [
              "Data 3",
            ]
          }
        >
          <td
            className="table-info"
            key="0"
          >
            Data 3
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly with no renderHeaders prop 1`] = `
<ListView
  data={
    Array [
      Array [
        "Ed",
      ],
      Array [
        "Edd",
      ],
      Array [
        "Eddie",
      ],
    ]
  }
  headerItems={
    Array [
      "Name",
    ]
  }
  renderHeaders={[Function]}
  renderRows={[Function]}
  tableClass="listview"
  tbodyClass="listview-tbody"
  tdClass="listview-td"
  thClass="listview-th"
  theadClass="listview-thead"
  trClass="listview-primary"
>
  <table
    className="listview"
  >
    <thead
      className="listview-thead"
    >
      <tr
        className="listview-primary"
      >
        <ElementMap
          HTMLTag="th"
          className="listview-td"
          items={
            Array [
              "Name",
            ]
          }
        >
          <th
            className="listview-td"
            key="0"
          >
            Name
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="listview-tbody"
    >
      <tr
        className="listview-primary"
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              "Ed",
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            Ed
          </td>
        </ElementMap>
      </tr>
      <tr
        className="listview-primary"
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              "Edd",
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            Edd
          </td>
        </ElementMap>
      </tr>
      <tr
        className="listview-primary"
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              "Eddie",
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            Eddie
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly with no renderHeaders prop 1`] = `
<ListView
  data={
    Array [
      Array [
        "Ed",
      ],
      Array [
        "Edd",
      ],
      Array [
        "Eddie",
      ],
    ]
  }
  headerItems={
    Array [
      "Name",
    ]
  }
  renderHeaders={[Function]}
  tableClass="listview"
  tbodyClass="listview-tbody"
  theadClass="listview-thead"
>
  <table
    className="listview"
  >
    <thead
      className="listview-thead"
    >
      <tr>
        <ElementMap
          HTMLTag="th"
          className="element-map"
          items={
            Array [
              "Name",
            ]
          }
        >
          <th
            className="element-map"
            key="0"
          >
            Name
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="listview-tbody"
    >
      <tr
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Ed",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Ed
          </td>
        </ElementMap>
      </tr>
      <tr
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Edd",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Edd
          </td>
        </ElementMap>
      </tr>
      <tr
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Eddie",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Eddie
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly with rows containing components 1`] = `
<ListView
  data={
    Array [
      Array [
        "Data 1",
      ],
      Array [
        <div>
          Data 2
        </div>,
      ],
      Array [
        <SomeComponent
          value="Data 3"
        />,
      ],
    ]
  }
  headerItems={
    Array [
      "Columns",
    ]
  }
  renderHeaders={[Function]}
  renderRows={[Function]}
  tableClass="table-striped"
  tbodyClass="table-active"
  tdClass="table-primary"
  thClass="table-info"
  theadClass="thead-dark"
  trClass="table-success"
>
  <table
    className="table-striped"
  >
    <thead
      className="thead-dark"
    >
      <tr
        className="table-success"
      >
        <ElementMap
          HTMLTag="th"
          className="table-primary"
          items={
            Array [
              "Columns",
            ]
          }
        >
          <th
            className="table-primary"
            key="0"
          >
            Columns
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="table-active"
    >
      <tr
        className="table-success"
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="table-info"
          items={
            Array [
              "Data 1",
            ]
          }
        >
          <td
            className="table-info"
            key="0"
          >
            Data 1
          </td>
        </ElementMap>
      </tr>
      <tr
        className="table-success"
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="table-info"
          items={
            Array [
              <div>
                Data 2
              </div>,
            ]
          }
        >
          <td
            className="table-info"
            key="0"
          >
            <div
              key="123"
            >
              Data 2
            </div>
          </td>
        </ElementMap>
      </tr>
      <tr
        className="table-success"
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="table-info"
          items={
            Array [
              <SomeComponent
                value="Data 3"
              />,
            ]
          }
        >
          <td
            className="table-info"
            key="0"
          >
            <SomeComponent
              key="456"
              value="Data 3"
            >
              <div>
                <span>
                  Data 3
                </span>
              </div>
            </SomeComponent>
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly without thead, tableClass, tbodyClass, tdClass, trClass, thClass 1`] = `
<ListView
  data={
    Array [
      Array [
        "Data 1",
      ],
      Array [
        999,
      ],
      Array [
        "Data 3",
      ],
    ]
  }
  renderHeaders={[Function]}
  renderRows={[Function]}
  tableClass="listview"
  tbodyClass="listview-tbody"
  tdClass="listview-td"
  thClass="listview-th"
  theadClass="listview-thead"
  trClass="listview-primary"
>
  <table
    className="listview"
  >
    <tbody
      className="listview-tbody"
    >
      <tr
        className="listview-primary"
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              "Data 1",
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            Data 1
          </td>
        </ElementMap>
      </tr>
      <tr
        className="listview-primary"
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              999,
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            999
          </td>
        </ElementMap>
      </tr>
      <tr
        className="listview-primary"
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="listview-th"
          items={
            Array [
              "Data 3",
            ]
          }
        >
          <td
            className="listview-th"
            key="0"
          >
            Data 3
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;
